File: 3399.smallest-substring-with-identical-characters-ii.py
#
# @lc app=leetcode id=3399 lang=python3
#
# [3399] Smallest Substring With Identical Characters II
#

# @lc code=start
class Solution:
    def minLength(self, s: str, numOps: int) -> int:
        def canAchieve(max_len):
            count_flips = 0
            zeroes = ones = 0
            left = 0
            for right in range(len(s)):
                if s[right] == '0':
                    zeroes += 1
                else:
                    ones += 1
                while min(zeroes, ones) > numOps:
                    if s[left] == '0':
                        zeroes -= 1
                    else:
                        ones -= 1
                    left += 1
                # Check if current window can be reduced within max_len by applying flips
                if (right - left + 1) - max(zeroes, ones) <= numOps and right - left + 1 <= max_len:
                    return True
            return False
        
        left, right = 1, len(s)
        while left < right:
            mid = (left + right) // 2
            if canAchieve(mid):
                right = mid
            else:
                left = mid + 1
        return left
# @lc code=end

File: 3399.smallest-substring-with-identical-characters-ii.java
#
# @lc app=leetcode id=3399 lang=java
#
# [3399] Smallest Substring With Identical Characters II
#

# @lc code=start
class Solution {
    public int minLength(String s, int numOps) {
        int n = s.length();
        int left = 1, right = n;
        int answer = n;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (canAchieve(s, numOps, mid)) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return answer;
    }
    // Check if we can make all runs of identical chars of length <= len with numOps flips
    private boolean canAchieve(String s, int numOps, int len) {
        // Try both: making all substrings at most len by flipping to '0' or '1'
        return minFlips(s, '0', len) <= numOps || minFlips(s, '1', len) <= numOps;
    }
    // For given char c, compute minimal flips to make all substrings of c of length <= len
    private int minFlips(String s, char c, int len) {
        int n = s.length();
        int flips = 0;
        int i = 0;
        while (i < n) {
            if (s.charAt(i) != c) {
                i++;
                continue;
            }
            int j = i;
            while (j < n && s.charAt(j) == c) j++;
            int runLen = j - i;
            if (runLen > len) {
                // Need to flip at least (runLen-1)/len substrings inside this run
                flips += (runLen - 1) / len;
            }
            i = j;
        }
        return flips;
    }
}
# @lc code=end